pkg/virt-handler/health/health.go

```diff
@@ -0,0 +1,112 @@
+package health
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"time"
+
+	v12 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"sync/atomic"
+
+	"k8s.io/apimachinery/pkg/util/runtime"
+
+	"kubevirt.io/kubevirt/pkg/api/v1"
+	"kubevirt.io/kubevirt/pkg/config"
+	"kubevirt.io/kubevirt/pkg/kubecli"
+	"kubevirt.io/kubevirt/pkg/log"
+	"kubevirt.io/kubevirt/pkg/virt-handler/device-manager"
+)
+
+var errorCount *uint64
+
+func init() {
+	runtime.ErrorHandlers = append(runtime.ErrorHandlers, func(err error) {
+		atomic.AddUint64(errorCount, 1)
+	})
+}
+
+func NewReadinesChecker(clientset kubecli.KubevirtClient, host string, clusterConfig *config.ClusterConfig) *ReadinesChecker {
+
+	return &ReadinesChecker{
+		clientset:     clientset,
+		host:          host,
+		clusterConfig: clusterConfig,
+	}
+}
+
+type ReadinesChecker struct {
+	clientset     kubecli.KubevirtClient
+	host          string
+	clusterConfig *config.ClusterConfig
+}
+
+// HeartBeat take a heartbeat inverval, a maximum of non-userfacing errors which are
+// allowed to happen and a stop channel to stop the heartbeat updates.
+// It periodically performs some health checks and updates the kubevirt.io/schedulable according to its checks.
+// Further it sets a timestamp on the node so that cluster components can see when it last updated the node.
+func (l *ReadinesChecker) HeartBeat(interval time.Duration, maxErrorsPerInterval uint64, stopCh chan struct{}) {
+	for {
+		wait.JitterUntil(func() {
+			schedulable := true
+
+			errorRateExceeded := atomic.LoadUint64(errorCount) > maxErrorsPerInterval
+			atomic.StoreUint64(errorCount, 0)
+
+			if errorRateExceeded {
+				// TODO we also need to try to patch the node with a reason
+				// TODO do we also want to panic? Maybe better to feed aliveness prove and decide on the manifest level
+				// if a restart it swanted.
+				schedulable = false
+			} else if useEmulation, _ := l.clusterConfig.IsUseEmulation(); !useEmulation {
+				tunErr := checkDeviceMissing(device_manager.TunPath, device_manager.TunName)
```

Comment:
Yes--for now. The tun device is a mandatory element of the current approach to networking. I assume that in the future this will become optional when networking is configurable.